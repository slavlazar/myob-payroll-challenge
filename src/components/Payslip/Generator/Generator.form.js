import React from 'react';
import PropTypes from 'prop-types';
import { Form, FastField } from 'formik';
import * as Yup from 'yup';

import { Input, Button, Columns, DateInput } from '../../shared/ui';

/**
 * Generator schema for this form
 * Contains validation for fields firstName, lastName, annualSalary, superRate
 * and payPeriod
 */
export const GeneratorSchema = Yup.object().shape({
  firstName: Yup.string().required('Please enter your First Name'),
  lastName: Yup.string().required('Please enter your Last Name'),
  annualSalary: Yup.number()
    .positive('Please enter a valid Annual Salary')
    .required('Please enter your Annual Salary'),
  superRate: Yup.number()
    .min(0, 'Please enter a valid Superannuation Rate (0 - 50%)')
    .max(50, 'Please enter a valid Superannuation Rate (0 - 50%)')
    .required('Please enter your Superannuation Rate'),
  payPeriod: Yup.date().required('Please provide a pay period'),
});

const currentDate = new Date();

/**
 * Generator form generated by Formik for our Payslip generate screen
 * @param {object} Params from Formik for this Payslip generator screen
 */
const GeneratorForm = ({ isSubmitting, handleSubmit, isValid }) => (
  <Form>
    <Columns>
      <FastField
        name="firstName"
        render={({ field, form }) => (
          <Input
            type="text"
            name="firstName"
            label="First Name"
            {...field}
            errorMessage={
              form.touched.firstName && form.errors.firstName
                ? form.errors.firstName
                : null
            }
          />
        )}
      />
      <FastField
        name="lastName"
        render={({ field, form }) => (
          <Input
            type="text"
            name="lastName"
            label="Last Name"
            {...field}
            errorMessage={
              form.touched.lastName && form.errors.lastName
                ? form.errors.lastName
                : null
            }
          />
        )}
      />
    </Columns>
    <Columns>
      <FastField
        name="annualSalary"
        render={({ field, form }) => (
          <Input
            type="number"
            name="annualSalary"
            label="Annual Salary"
            {...field}
            errorMessage={
              form.touched.annualSalary && form.errors.annualSalary
                ? form.errors.annualSalary
                : null
            }
          />
        )}
      />
      <FastField
        name="superRate"
        render={({ field, form }) => (
          <Input
            type="number"
            name="superRate"
            label="Superannuation Rate"
            {...field}
            errorMessage={
              form.touched.superRate && form.errors.superRate
                ? form.errors.superRate
                : null
            }
          />
        )}
      />
      {/* 
        Unfortunately react datepicker currently does not honour the 
        maxDate prop for month picker.
        This was fixed in this commit below - so the payslip month selection 
        currently lets you go over the current month :(
        https://github.com/gautam-pahuja/react-datepicker/commit/66c1e370b6cb24ad7aef86fc384067159aff43ab 
      */}
      <FastField
        name="payPeriod"
        render={({ field, form }) => (
          <DateInput
            name="payPeriod"
            label="Pay Period"
            todayButton="This month"
            showMonthYearPicker
            maxDate={currentDate}
            {...field}
            formik={{ ...form }}
            errorMessage={
              form.touched.payPeriod && form.errors.payPeriod
                ? form.errors.payPeriod
                : null
            }
          />
        )}
      />
    </Columns>
    <Button
      id="paylsip-generate-button"
      type="button"
      colour="primary"
      size="small"
      onClick={handleSubmit}
      disabled={!isValid || isSubmitting}
      aria-label="Generate Payslip"
    >
      Generate Payslip
    </Button>
  </Form>
);

GeneratorForm.propTypes = {
  /** Values from Formik */
  values: PropTypes.object.isRequired,
  /** Object of touched elements from Formik */
  touched: PropTypes.object.isRequired,
  /** Function to handle form submission */
  handleSubmit: PropTypes.func.isRequired,
  /** Function to handle field change */
  handleChange: PropTypes.func.isRequired,
  /** Function to handle field blur */
  handleBlur: PropTypes.func.isRequired,
  /** Boolean to determine if form passes validation schema */
  isValid: PropTypes.bool.isRequired,
  isSubmitting: PropTypes.bool,
  errors: PropTypes.object,
};

GeneratorForm.defaultProps = {
  isSubmitting: false,
  errors: {},
};

export default GeneratorForm;
