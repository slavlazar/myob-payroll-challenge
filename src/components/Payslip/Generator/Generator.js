import React from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';

import { Alert, Card } from '../../shared/ui';
import GeneratorForm, { GeneratorSchema } from './Generator.form';
import { generateInitialValuesFromSchema } from '../../../utils/helpers';

/**
 * Payslip Generator component displaying our view with payslip generate form
 * generated by Formik
 * @param {object} Generator component params from our Payslip state
 */
const Generator = ({ handleSubmit, loading, error }) => (
  <Card loading={loading}>
    {error && (
      <Alert
        className="payslip-generator-error"
        type="danger"
        title="Error!"
        size="small"
      >
        {error}
      </Alert>
    )}
    <Formik
      component={GeneratorForm}
      onSubmit={handleSubmit}
      validationSchema={GeneratorSchema}
      initialValues={generateInitialValuesFromSchema(
        Object.keys(GeneratorSchema.fields)
      )}
    />
  </Card>
);

Generator.propTypes = {
  /** Function to handle payslip generation */
  handleSubmit: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  error: PropTypes.string,
};

Generator.defaultProps = {
  loading: false,
  error: null,
};

export default Generator;
