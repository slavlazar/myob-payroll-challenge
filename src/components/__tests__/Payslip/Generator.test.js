import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import Generator from '../../Payslip/Generator/Generator';
import GeneratorForm from '../../Payslip/Generator/Generator.form';

const handleSubmit = sinon.spy();

const shallowWrapper = (props = {}) => {
  const allProps = {
    handleSubmit,
    ...props,
  };

  return shallow(<Generator {...allProps} />);
};

const defaultFormProps = {
  values: {},
  touched: {},
  handleSubmit,
  handleChange: sinon.spy(),
  handleBlur: sinon.spy(),
  isValid: false,
};

const shallowGeneratorFormWrapper = (props = {}) => {
  const allProps = {
    ...defaultFormProps,
    ...props,
  };

  return shallow(<GeneratorForm {...allProps} />);
};

describe('Components: Payslip: Generator', () => {
  describe('Generator', () => {
    it('renders correctly', () => {
      const wrapped = shallowWrapper();
      expect(wrapped).toMatchSnapshot();
    });

    it('should render an alert message', () => {
      const wrapped = shallowWrapper({ error: 'Error' });
      expect(wrapped.find('.payslip-generator-error')).toHaveLength(1);
    });
  });

  describe('Generator Form', () => {
    afterEach(() => {
      defaultFormProps.handleSubmit.resetHistory();
      defaultFormProps.handleChange.resetHistory();
      defaultFormProps.handleBlur.resetHistory();
    });

    it('renders correctly', () => {
      const wrapped = shallowGeneratorFormWrapper();

      expect(wrapped).toMatchSnapshot();
    });

    it('should generate the required fields', () => {
      const wrapped = shallowGeneratorFormWrapper();
      expect(wrapped.find('[name="firstName"]')).toHaveLength(1);
      expect(wrapped.find('[name="lastName"]')).toHaveLength(1);
      expect(wrapped.find('[name="annualSalary"]')).toHaveLength(1);
      expect(wrapped.find('[name="superRate"]')).toHaveLength(1);
      expect(wrapped.find('[name="payPeriod"]')).toHaveLength(1);
    });

    it('should trigger the onSubmit event', () => {
      const wrapped = shallowGeneratorFormWrapper();

      expect(defaultFormProps.handleSubmit.callCount).toEqual(0);

      const button = wrapped.find('#paylsip-generate-button');
      button.simulate('click');
      expect(defaultFormProps.handleSubmit.callCount).toEqual(1);
    });
  });
});
