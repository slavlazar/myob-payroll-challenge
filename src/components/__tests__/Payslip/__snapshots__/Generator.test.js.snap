// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: Payslip: Generator Generator Form renders correctly 1`] = `
<Form>
  <styled.div>
    <FormikConnect(FastFieldInner)
      name="firstName"
      render={[Function]}
    />
    <FormikConnect(FastFieldInner)
      name="lastName"
      render={[Function]}
    />
  </styled.div>
  <styled.div>
    <FormikConnect(FastFieldInner)
      name="annualSalary"
      render={[Function]}
    />
    <FormikConnect(FastFieldInner)
      name="superRate"
      render={[Function]}
    />
    <FormikConnect(FastFieldInner)
      name="payPeriod"
      render={[Function]}
    />
  </styled.div>
  <Button
    aria-label="Generate Payslip"
    block={false}
    colour="primary"
    disabled={true}
    id="paylsip-generate-button"
    onClick={[Function]}
    size="small"
    type="button"
  >
    Generate Payslip
  </Button>
</Form>
`;

exports[`Components: Payslip: Generator Generator renders correctly 1`] = `
<Card
  loading={false}
>
  <Formik
    component={[Function]}
    enableReinitialize={false}
    initialValues={
      Object {
        "annualSalary": "",
        "firstName": "",
        "lastName": "",
        "payPeriod": "",
        "superRate": "",
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
    validationSchema={
      ObjectSchema {
        "_blacklist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "_conditions": Array [],
        "_defaultDefault": [Function],
        "_deps": Array [],
        "_excludedEdges": Array [],
        "_exclusive": Object {},
        "_mutate": undefined,
        "_nodes": Array [
          "payPeriod",
          "superRate",
          "annualSalary",
          "lastName",
          "firstName",
        ],
        "_options": Object {
          "abortEarly": true,
          "recursive": true,
        },
        "_type": "object",
        "_typeError": [Function],
        "_whitelist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "fields": Object {
          "annualSalary": NumberSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "min": true,
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "number",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "firstName": StringSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "string",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "lastName": StringSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "string",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "payPeriod": DateSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "date",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "superRate": NumberSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "max": true,
              "min": true,
              "required": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "number",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
        },
        "tests": Array [],
        "transforms": Array [
          [Function],
        ],
      }
    }
  />
</Card>
`;
